GH-2280: Fix misleading host functions gas charges

mpapierski opened the pull request:
Closes #2280

This PR fixes a problem where `transfer_to_account` and `transfer_from_purse_to_account` (and also `create_purse`) host functions are charging not only for host function cost, but also for a system contract charge.

Using those APIs leads to a situation where the user who uses the absolute minimum payment amount to cover the host function cost will receive a `GasLimit` error. In this case, the system will also incorrectly attempt to charge for system contracts.

Commit:
Add regression tests for the issue.

Commit:
Fix misleading system contract charges

Commit:
Increase transfer host functions.

This is done to make those functions cost always the amount required to
create new account.

luis-casper commented on execution_engine/src/shared/scoped_guard.rs:
```diff
@@ -0,0 +1,49 @@
+impl ScopedCountingGuard {
+    #[must_use]
+    pub(crate) fn new(shared_value: &Rc<RefCell<i32>>) -> Self {
+        shared_value.replace_with(|&mut value| value + 1);
```
Why do we use `replace_with` here instead of `borrow_mut` as in L20?

mpapierski replied:
Addressed

luis-casper commented on execution_engine/src/shared/scoped_guard.rs:
```diff
@@ -0,0 +1,49 @@
+impl ScopedCountingGuard {
+    #[must_use]
+    pub(crate) fn new(shared_value: &Rc<RefCell<i32>>) -> Self {
+        shared_value.replace_with(|&mut value| value + 1);
```
We might also want to use a checked addition here, even if we panic on overflow.

mpapierski replied:
I addressed this. After thinking, I decided not to panic but have another error in case of overflow. I don't see any scenario where it would hit this error other than a programming error.

luis-casper replied:
But then I think we should panic rather than having the error variant, as it would be a programming error.

luis-casper commented on execution_engine_testing/tests/src/test/regression/gh_2280.rs:
```diff
@@ -0,0 +1,359 @@
+
+#[ignore]
+#[test]
+fn gh_2280_create_purse_should_always_cost_the_same_gas() {
```
This function is largely repeated code from the function above.
We should refactor these if possible.

mpapierski replied:
Refactored as much of duplicated code as possible out of it

luis-casper commented on execution_engine/src/core/runtime/mod.rs:
```diff
@@ -64,6 +67,9 @@ pub struct Runtime<'a, R> {
     call_stack: Vec<CallStackElement>,
+    /// Value larger than 0 indicates executing under host function context, so we don't need to
+    /// charge for system contracts.
+    host_function_counter: Rc<RefCell<i32>>,
```
We might want to consider giving this type a name.

mpapierski replied:
Addressed

Commit:
Apply review comments form Luis

Commit:
Refactor as much testing code as possible

Commit:
Add extra tests for stored contracts.

Commit:
Fix rustdoc

luis-casper commented on execution_engine/src/shared/scoped_guard.rs:
```diff
@@ -1,23 +1,75 @@
+        *self.0.borrow()
+    }
+
+    pub(crate) fn is_set(&self) -> bool {
```
Maybe we should name this `is_zero` or `is_positive`, since counters can be set to zero.

luis-casper commented on execution_engine/src/shared/scoped_guard.rs:
```diff
@@ -27,23 +79,36 @@ mod tests {
+
         {
-            assert_eq!(*value.borrow(), 0);
+            assert_eq!(counter.count(), 0);
```
The `is_*` method could be used.

mpapierski replied:
Addressed

luis-casper commented on execution_engine/src/core/execution/error.rs:
```diff
@@ -156,6 +156,9 @@ pub enum Error {
     /// Missing system contract hash.
     #[error("Missing system contract hash: {0}")]
     MissingSystemContractHash(String),
+    /// Host function reached depth limit.
```
Since this limit is `2^64-1`, it doesn't make sense to have this error variant.
Reaching the limit (with the system contracts only) is clearly a bug and panicking would be preferable.

mpapierski replied:
Fair - I removed this error variant, restored `#[must_use]` and added a `// SAFETY:` comment around `.expect()` call.

Commit:
Address Luis feedback

This makes host function overflow a programming error rather than a
separate error variant

Fraser999 committed:
make scoped_guard more specific

mpapierski committed:
Merge pull request #3 from Fraser999/gh-2280-host-functions-gas

make scoped_guard more specific

Commit:
Help detect this programming error in development.

mpapierski commented:
bors r+
